fix: initial value in FAST select control not used when using Angular two-way binding

brianehenry opened the pull request:
<!---
Thanks for filing a pull request 😄 ! Before you submit, please read the following:

Search open/closed issues before submitting. Someone may have pushed the same thing before!

Provide a summary of your changes in the title field above.
-->

# Pull Request

## 📖 Description

<!---
Provide some background and a description of your work.
What problem does this change solve?
Is this a breaking change, chore, fix, feature, etc?
-->

The FAST select control fails to accept its initial value when it is set right after the select is connected. This change handles that case by marking the `option` element corresponding to the given value with the `selected` attribute. 

### 🎫 Issues

<!---
* List and link relevant issues here.
-->

Fixes #5139 

## 👩‍💻 Reviewer Notes

<!---
Provide some notes for reviewers to help them provide targeted feedback and testing.

Do you recommend a smoke test for this PR? What steps should be followed?
Are there particular areas of the code the reviewer should focus on?
-->

## 📑 Test Plan

<!---
Please provide a summary of the tests affected by this work and any unique strategies employed in testing the features/fixes.
-->

Added one unit test that reflects the desired behavior.

## ✅ Checklist

### General

<!--- Review the list and put an x in the boxes that apply. -->

- [x] I have included a change request file using `$ yarn change`
- [x] I have added tests for my changes.
- [x] I have tested my changes.
- [ ] I have updated the project documentation to reflect my changes.
- [x] I have read the [CONTRIBUTING](https://github.com/Microsoft/fast/blob/master/CONTRIBUTING.md) documentation and followed the [standards](https://www.fast.design/docs/community/code-of-conduct/#our-standards) for this project.

### Component-specific

<!--- Review the list and put an x in the boxes that apply. -->
<!--- Remove this section if not applicable. -->

- [ ] I have added a new component
- [ ] I have modified an existing component
- [ ] I have updated the [definition file](https://github.com/Microsoft/fast/blob/master/packages/web-components/fast-components/CONTRIBUTING.md#definition)
- [ ] I have updated the [configuration file](https://github.com/Microsoft/fast/blob/master/packages/web-components/fast-components/CONTRIBUTING.md#configuration)

## ⏭ Next Steps

<!---
If there is relevant follow-up work to this PR, please list any existing issues or provide brief descriptions of what you would like to do next.
-->

radium-v reviewed (changes requested):
Hi @brianehenry, thanks for investigating this issue. I downloaded and ran the angular demo from #5139 to make sure my review is accurate.

The approach works for setting the default selected option when `ngModel` is used, but there are two problems:
* It sets the `selected` attribute on the matching `fast-option`, which can cause collisions when both a default `value` exists and an initial child option already has the `selected` attribute.
* It runs the check every time `value` is updated.

Since we only need it to check this when the component is connected, we can override the `setDefaultSelectedOption()` method for `Select` in the same way that `Combobox` does it.

This override function is a modified version from `Listbox`, which allows the options list to fallback to the filtered list of child elements, and adds a `value` check for each option:

```ts
    protected setDefaultSelectedOption(): void {
        if (this.$fastController.isConnected) {
            const options: ListboxOption[] =
                this.options ??
                Array.from(this.children).filter(Listbox.slottedOptionFilter);

            const selectedIndex = options?.findIndex(
                el =>
                    el.getAttribute("selected") !== null ||
                    el.selected ||
                    el.value === this.value
            );

            if (selectedIndex !== -1) {
                this.selectedIndex = selectedIndex;
                return;
            }

            this.selectedIndex = 0;
        }
    }
```

Then to make sure the `formResetCallback` functions as expected, we can change the method to call the super method:

```ts
public formResetCallback = (): void => {
    this.setProxyOptions();
    super.setDefaultSelectedOption();
    this.value = this.firstSelectedOption.value;
};
```

brianehenry commented:
Hi @radium-v, thanks for the detailed feedback. 

I tried out your suggested changes, but they don't seem to fix the issue that I'm seeing (or, at least, the test case I wrote fails with those changes). The issue with your suggestion is that it relies on the select's private `_value` field being set by the `value` setter. But the `value` setter does not set `_value` if it does not find a matching `option` element with the same `value` string, which is the case when the `options` property has not yet been initialized. So, when `setDefaultSelectedOption()` is called, the select's `value` property is still unset.

I updated my PR to address the first issue you had with my changes - the select will now clear any previously set `selected` attribute on the option elements. I don't think the second issue you raised is a showstopper, as the `value` setter won't call any of the code I added after the `options` property is initialized.

EisenbergEffect commented:
@radium-v Can you take a look at this one again. Brian has made some changes since your last feedback.

brianehenry commented:
@radium-v Provided some additional feedback [on discord](https://discord.com/channels/449251231706251264/449251231706251266/931692250679873547). I applied the suggested changes to my branch.

EisenbergEffect reviewed (approved):
Just need a bit of code commenting for the on out-of-the-ordinary situation noted below.

radium-v commented on packages/web-components/fast-foundation/src/select/select.ts:
```diff
@@ -305,6 +306,25 @@ export class Select extends FormAssociatedSelect {
+
+        const selectedIndex = options?.findIndex(
+            el =>
+                el.getAttribute("selected") !== null ||
```
My bad, there's a better check for this:

```suggestion
                el.hasAttribute("selected") ||
```

brianehenry commented on packages/web-components/fast-foundation/src/select/select.ts:
```diff
@@ -229,7 +230,7 @@ export class Select extends FormAssociatedSelect {
     public formResetCallback(): void {
         this.setProxyOptions();
-        this.setDefaultSelectedOption();
+        super.setDefaultSelectedOption();
```
Added a comment with an explanation. For what it's worth, I find it strange to call a `super` method from anywhere besides its override, but I don't see an obvious alternative here.

brianehenry commented:
@chrisdholt @EisenbergEffect This PR is ready to be merged.
